### PR TITLE
XS-3032: Logic for 0041 was flawed, corrected.

### PR DESCRIPTION
**Pull request by andrewperkins-wf**
# ULTIMATE PROBLEM
More issues with the new rules that we needed to get fixed.  Mostly around None checks and the like.

# NOTE
# HOW IT WAS FIXED
* None checks, value checks, and more!

# TESTING SUGGESTIONS
# AREAS AFFECTED
# UNIT TESTS

### CODE REVIEW

File: dqc_us_rules/dqc_us_0013.py [hunk 1]
```diff
@@ -109,12 +108,13 @@ def filter_negative_number_with_dependence_facts(val, blacklist_concepts):
...
-            f.concept.type is not None and
+            fact for fact in numeric_facts
+            if fact.xValue < 0 and
+            fact.concept.type is not None and
             # facts with numerical values less than 0 and contexts and
-            f.context is not None and
+            fact.context is not None
             # check xsd type of the concept
-            f.isNumeric
+            # fact.isNumeric
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0013.py]: Commented code left in place?

File: dqc_us_rules/dqc_us_0014.py [hunk 2]
```diff
@@ -68,16 +67,17 @@ def filter_negative_number_no_dimensions_facts(val, blacklist_concepts):
...
         fact.concept.type is not None and
         # facts with numerical values less than 0 (negative) and contexts
         fact.context is not None and
         fact.context.segDimValues is not None and
         # check that the fact does not have dimensions
-        len(fact.context.segDimValues) == 0 and
+        len(fact.context.segDimValues) == 0
         # check xsd type of the concept
-        fact.isNumeric
+        # fact.isNumeric
```
Comment (jasonleinbach-wf) [inline dqc_us_rules/dqc_us_0014.py]: Commented?

**Comment by jasonleinbach-wf** [comment]
+1 pending nits.

**Comment by andrewperkins-wf** [comment]
+1

**Comment by jasonleinbach-wf** [comment]
+1

**Comment by chrislococo-wf** [comment]
+1 (for what I didn't code)

**Comment by jasonleinbach-wf** [comment]
+1

**Comment by brianmyers-wf** [comment]
+1

**Comment by jasonleinbach-wf** [comment]
+1

**Comment by brianmyers-wf** [comment]
flake8 produced no warnings 
and unit tests all passed:
```
$ flake8 dqc_us_rules
$ ./run-unit-tests.sh
Running: nosetests tests
Verbosity: 2

consortium_rules on: refs/heads/XS-3032
consortium_rules at: 15d597cb72a21cf97907dcc968ca57e8131733a9

Tests axis_qnames with valid axis ... ok
Tests axis_qnames with no axis ... ok
Tests get_facts_dei ... ok
Tests get_facts_with_type ... ok
Tests the lookup_dei_facts ... ok
Tests to make sure that a Fact with a None type context is not valid ... ok
Tests to make sure that a None type fact is not valid ... ok
Tests to make sure that a Fact.context with a None type segDimValue is ... ok
Tests to make sure that a valid fact still works ... ok
Tests to see if duplicate facts with a contextID equal to None are ... ok
Tests to see if duplicate facts with the same contextID are filtered ... ok
Test to see if ignore units works when filtering duplicate facts ... ok
Test to see if isNil filtering correctly works ... ok
Tests to see if xValid filtering works correctly ... ok
Tests lookup_gaap_facts ... ok
Tests member_qnames with valid member_qnames ... ok
Tests prepare_for_calculation when the first item needs to be ignored ... ok
Tests prepare_facts_for_calculation when ignore units is on ... ok
Tests prepare_facts_for_calculation when ignore units is on but nothing ... ok
Tests prepare_facts_for_calculation when the last item needs to be ... ok
Tests to see prepare_facts_for_calculation works correctly when there ... ok
Tests to see if prepare_facts_for_calculation works when the path is ... ok
Tests scaling all facts by infinity ... ok
Tests scaling without any scale values ... ok
Tests scaling all values by the same value ... ok
Test scaling with values that can't be scaled ... ok
Tests scaling some facts by infinity that are not in order ... ok
Tests scaling by negative numbers ... ok
Tests scaling with possitive numbers ... ok
Tests scaling some facts by infinity ... ok
Tests scaling with one of the scale values missing ... ok
Tests to see if no errors are throws by a none context ... ok
Tests to see if errors right numher of errors are thrown ... ok
Tests to see if no errors are thrown without a duration on fact ... ok
Make sure that the scale value is valid if it is infinity ... ok
Make sure that the scale value is valid when it is zero ... ok
Make sure that the scale value is not valid if it is Not a Number ... ok
Make sure that the scale value is valid if it is negative infinity ... ok
Make sure that the scale value is not valid if is not None ... ok
Tests values_unequal with equal values ... ok
Tests values_unequal with values that are calculated to be equal ... ok
Test values_unequal with unequal values ... ok
Test axis_exists on one fact where the axis exists and one fact where ... ok
Tests _dei_pattern against a pass list and a fail list ... ok
Test _get_end_of_period with no concept ... ok
test_run_checks_EntityCommonStockSharesOutstanding (tests.unit_tests.test_dqc_us_0005.TestContextChecks) ... ok
test_run_checks_axis_exists (tests.unit_tests.test_dqc_us_0005.TestContextChecks) ... ok
test_run_checks_axis_member_exists (tests.unit_tests.test_dqc_us_0005.TestContextChecks) ... ok
Test _dei_pattern against pass list and fail list ... ok
Test _dict_list_update ... ok
Test to make sure that the dictionary read in from the csv shares ... ok
Test to make sure that the dictionary read is doesn't equal something ... ok
Tests _compare_facts with empty facts ... ok
Tests compare facts with different values and one value that is the ... ok
Verifies the check for negative numbers with dependencies works as ... ok
Verifies that the check for whether a precondition concept exists ... ok
test_grab_numeric_facts (tests.unit_tests.test_dqc_us_0014.TestDQC0014) ... ok
test_negative_number_no_dimensions (tests.unit_tests.test_dqc_us_0014.TestDQC0014) ... ok
Tests blacklist to see if it contains the right items and is the right ... ok
Verifies the actual exclusion rule config can be parsed with no ... ok
Test contains with value that is in list and value that isn't in list ... ok
Test contains_ignore_case with item that contain the same letters but ... ok
Test equals with values that should be equal and values that aren't ... ok
Test _parse_row against expected result ... ok
Tests that fact errors if it contains a deprecated concept ... ok
Tests that fact errors if it contains a deprecated dimension ... ok
Tests to make sure that fact errors it it contains a deprecated member ... ok
Tests that fact is checkable ... ok
Tests _doc_period_end_date_check to see of the length is right and that ... ok
Tests _doc_period_end_date_check when it should return a warning and an ... ok
Tests _doc_period_end_date_check when it should return an error ... ok
Tests _setup_facts against expected dped and dei dicts ... ok
Tests to make sure that _get_default_dped on a dict with a None and a ... ok
Tests to make sure that _get_default_dped return the first non Nil ... ok
Tests to make sure that _get_default_dped on an empty dict returns None ... ok
Tests to make sure that _is_valid_eop_fact catches errors ... ok
Tests to see if dqc_us_0041_errors will be caught ... ok
Tests to make sure dqc_us_0041 doesn't always throw errors ... ok
Tests to see if _default_dimension_mismatch can determine if two ... ok
Tests to make sure the _default_dimension_mismatch doesn't always throw ... ok
Tests handling of the validation not having a usgaapDefaultDimensions ... ok
Tests to make sure that init.py works correctly ... ok

----------------------------------------------------------------------
Ran 82 tests in 0.921s

OK
```
Started local arelle gui and successfully pushed a test through with DQC plugin loaded and no exceptions or unexpected errors.

**Comment by andrewperkins-wf** [comment]
Reviewed and unit tested.  Merging to get another release-candidate build released.

**Pull request merged**